File: app/src/main/java/com/krystianwsul/checkme/persistencemodel/DailyScheduleRecord.java
package com.krystianwsul.checkme.persistencemodel;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.List;

public class DailyScheduleRecord extends Record {
    static final String TABLE_DAILY_SCHEDULES = "dailySchedules";

    private static final String COLUMN_SCHEDULE_ID = "scheduleId";
    private static final String COLUMN_CUSTOM_TIME_ID = "customTimeId";
    private static final String COLUMN_HOUR = "hour";
    private static final String COLUMN_MINUTE = "minute";

    private final int mScheduleId;

    private final Integer mCustomTimeId;

    private final Integer mHour;
    private final Integer mMinute;

    public static void onCreate(@NonNull SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.execSQL("CREATE TABLE " + TABLE_DAILY_SCHEDULES
                + " (" + COLUMN_SCHEDULE_ID + " INTEGER NOT NULL UNIQUE REFERENCES " + ScheduleRecord.TABLE_SCHEDULES + "(" + ScheduleRecord.COLUMN_ID + "), "
                + COLUMN_CUSTOM_TIME_ID + " INTEGER REFERENCES " + LocalCustomTimeRecord.Companion.getTABLE_CUSTOM_TIMES() + "(" + LocalCustomTimeRecord.Companion.getCOLUMN_ID() + "), "
                + COLUMN_HOUR + " INTEGER, "
                + COLUMN_MINUTE + " INTEGER);");
    }

    @NonNull
    static List<DailyScheduleRecord> getDailyScheduleRecords(@NonNull SQLiteDatabase sqLiteDatabase) {
        List<DailyScheduleRecord> dailyScheduleTimeRecords = new ArrayList<>();

        Cursor cursor = sqLiteDatabase.query(TABLE_DAILY_SCHEDULES, null, null, null, null, null, null);
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            dailyScheduleTimeRecords.add(cursorToDailyScheduleTimeRecord(cursor));
            cursor.moveToNext();
        }
        cursor.close();

        return dailyScheduleTimeRecords;
    }

    @NonNull
    private static DailyScheduleRecord cursorToDailyScheduleTimeRecord(@NonNull Cursor cursor) {
        int scheduleId = cursor.getInt(0);
        Integer customTimeId = (cursor.isNull(1) ? null : cursor.getInt(1));
        Integer hour = (cursor.isNull(2) ? null : cursor.getInt(2));
        Integer minute = (cursor.isNull(3) ? null : cursor.getInt(3));

        Assert.assertTrue((hour == null) == (minute == null));
        Assert.assertTrue((hour == null) || (customTimeId == null));
        Assert.assertTrue((hour != null) || (customTimeId != null));

        return new DailyScheduleRecord(true, scheduleId, customTimeId, hour, minute);
    }

    DailyScheduleRecord(boolean created, int scheduleId, @Nullable Integer customTimeId, @Nullable Integer hour, @Nullable Integer minute) {
        super(created);

        Assert.assertTrue((hour == null) == (minute == null));
        Assert.assertTrue((hour == null) || (customTimeId == null));
        Assert.assertTrue((hour != null) || (customTimeId != null));

        mScheduleId = scheduleId;

        mCustomTimeId = customTimeId;

        mHour = hour;
        mMinute = minute;
    }

    int getScheduleId() {
        return mScheduleId;
    }

    @Nullable
    public Integer getCustomTimeId() {
        return mCustomTimeId;
    }

    @Nullable
    public Integer getHour() {
        return mHour;
    }

    @Nullable
    public Integer getMinute() {
        return mMinute;
    }

    @NonNull
    @Override
    ContentValues getContentValues() {
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_SCHEDULE_ID, mScheduleId);
        contentValues.put(COLUMN_CUSTOM_TIME_ID, mCustomTimeId);
        contentValues.put(COLUMN_HOUR, mHour);
        contentValues.put(COLUMN_MINUTE, mMinute);
        return contentValues;
    }

    @NonNull
    @Override
    UpdateCommand getUpdateCommand() {
        return getUpdateCommand(TABLE_DAILY_SCHEDULES, COLUMN_SCHEDULE_ID, mScheduleId);
    }

    @NonNull
    @Override
    InsertCommand getInsertCommand() {
        return getInsertCommand(TABLE_DAILY_SCHEDULES);
    }

    @NonNull
    @Override
    DeleteCommand getDeleteCommand() {
        return getDeleteCommand(TABLE_DAILY_SCHEDULES, COLUMN_SCHEDULE_ID, mScheduleId);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/persistencemodel/DailyScheduleRecord.kt
package com.krystianwsul.checkme.persistencemodel

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

class DailyScheduleRecord(created: Boolean, val scheduleId: Int, val customTimeId: Int?, val hour: Int?, val minute: Int?) : Record(created) {

    companion object {

        val TABLE_DAILY_SCHEDULES = "dailySchedules"

        private val COLUMN_SCHEDULE_ID = "scheduleId"
        private val COLUMN_CUSTOM_TIME_ID = "customTimeId"
        private val COLUMN_HOUR = "hour"
        private val COLUMN_MINUTE = "minute"

        fun onCreate(sqLiteDatabase: SQLiteDatabase) {
            sqLiteDatabase.execSQL("CREATE TABLE $TABLE_DAILY_SCHEDULES " +
                    "($COLUMN_SCHEDULE_ID INTEGER NOT NULL UNIQUE REFERENCES ${ScheduleRecord.TABLE_SCHEDULES} (${ScheduleRecord.COLUMN_ID}), " +
                    "$COLUMN_CUSTOM_TIME_ID INTEGER REFERENCES ${LocalCustomTimeRecord.TABLE_CUSTOM_TIMES} (${LocalCustomTimeRecord.COLUMN_ID}), " +
                    "$COLUMN_HOUR INTEGER, " +
                    "$COLUMN_MINUTE INTEGER);")
        }

        fun getDailyScheduleRecords(sqLiteDatabase: SQLiteDatabase) = getRecords(sqLiteDatabase, TABLE_DAILY_SCHEDULES, this::cursorToDailyScheduleTimeRecord)

        private fun cursorToDailyScheduleTimeRecord(cursor: Cursor) = cursor.run {
            val scheduleId = getInt(0)
            val customTimeId = if (isNull(1)) null else getInt(1)
            val hour = if (isNull(2)) null else getInt(2)
            val minute = if (isNull(3)) null else getInt(3)

            check(hour == null == (minute == null))
            check(hour == null || customTimeId == null)
            check(hour != null || customTimeId != null)

            DailyScheduleRecord(true, scheduleId, customTimeId, hour, minute)
        }
    }

    init {
        check(hour == null == (minute == null))
        check(hour == null || customTimeId == null)
        check(hour != null || customTimeId != null)
    }

    override val contentValues
        get() = ContentValues().apply {
        put(COLUMN_SCHEDULE_ID, scheduleId)
        put(COLUMN_CUSTOM_TIME_ID, customTimeId)
        put(COLUMN_HOUR, hour)
        put(COLUMN_MINUTE, minute)
    }

    override val commandTable = TABLE_DAILY_SCHEDULES
    override val commandIdColumn = COLUMN_SCHEDULE_ID
    override val commandId = scheduleId
}
